### PR TITLE
feat(source): add source preview surfaces

### PR DESCRIPTION
**Pull request by jonathanhaaswriter**
## Summary
- add preview-oriented source operations for check, discover, and read
- expose those operations over HTTP and Connect on the bootstrap app
- add a `cerebro source` CLI so the GitHub fixture source is usable end-to-end locally

## Testing
- make verify
- ./bin/cerebro source list
- ./bin/cerebro source check github token=test
- ./bin/cerebro source discover github token=test
- ./bin/cerebro source read github token=test

### CODE REVIEW

**Comment by jonathanhaaswriter** [comment]
@droid Please perform a deep code review and security review of this PR. Leave inline comments only for high-confidence, actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review this PR after the latest fixes. Focus on whether the prior code review and security findings are resolved, and call out any remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
### Droid sweep status

Latest sweep: `2026-04-29 22:09 UTC`.

- Unresolved non-outdated review threads: `0`
- CI `verify`: `SUCCESS`
- `droid-review`: `not present`

I’ll keep watching for new review findings/check failures and address them as they appear.

**Comment by jonathanhaaswriter** [comment]
Re-audited the live unresolved thread (including outdated) against the current branch tip. `sourceConfigFromQuery` now rejects sensitive config keys in query params and still accepts bearer-token auth via headers, so the secret-leak concern is already addressed on-branch; no code change was needed. Validation run: none locally during this audit; existing branch `verify` is green. Remaining unresolved threads after this audit: 0.

**Comment by jonathanhaaswriter** [comment]
Droid sweep audit complete (final pass)

- Re-scanned all live review threads, including outdated ones
- Verified each remaining thread against the current branch tip
- Resolved every actionable thread (or fixed it inline and pushed)
- 0 unresolved review threads across the whole open-PR stack at this time

PR #494 (fix/feedback-main-followup) and PR #495 (feat/droid-creation-workflows) received additional fixes for SSRF (DNS-rebind) and workflow branch-name validation in this pass.

**Comment by jonathanhaaswriter** [comment]
@droid review

**Comment by jonathanhaaswriter** [comment]
@droid please harden this PR end-to-end:

1. Re-review the diff for residual bugs, races, edge cases, security gaps, and untested paths.
2. Apply the smallest correct fixes for any high-confidence issues you find.
3. Strengthen test coverage: add regression tests for every prior review finding on this PR plus the new behavior, including failure paths, boundary inputs, concurrency, and not-found / typed-nil interface guards.
4. Run focused tests and `make verify`. If validators fail, fix until green.
5. Commit and push the fixes to this PR's branch (or, if you lack push permission in this workflow, leave concrete inline suggestions and a summary comment).

Do not change the PR scope, do not modify the default branch, and do not merge.

**Comment by jonathanhaaswriter** [comment]
@droid review

File: internal/bootstrap/app.go [hunk 1]
```diff
@@ -143,6 +224,64 @@ func writeProtoJSON(w http.ResponseWriter, statusCode int, message proto.Message
...
+}
+
+func sourceConnectError(err error) error {
+	if errors.Is(err, sourceops.ErrInvalidSourceID) {
+		return connect.NewError(connect.CodeInvalidArgument, err)
+	}
+	if errors.Is(err, sourceops.ErrSourceNotFound) {
+		return connect.NewError(connect.CodeNotFound, err)
+	}
+	return err
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/bootstrap/app.go:274]: Fixed in 59c4863ce by wrapping source validation failures with `ErrInvalidSourceConfig` and mapping that sentinel to Connect `CodeInvalidArgument`; tests cover missing token and invalid cursor.

**Comment by jonathanhaaswriter** [comment]
@droid review

Please re-review the source preview surfaces deeply. Focus on the new public surface and how it interacts with untrusted upstream data:

1. **Input validation and abuse**: verify all preview parameters (per_page, owner/repo/scope, family, cursors) are bounded and rejected as `InvalidArgument` when malformed, with no path that lets an attacker request unbounded pages or pivot the source to arbitrary URLs.
2. **Tenant/runtime scoping**: ensure preview reads are constrained to the caller’s tenant/runtime context and cannot be used to enumerate sources or configs belonging to other tenants.
3. **Secret handling**: confirm tokens, query-string secrets, basic-auth headers, and config blobs never appear in logs, error messages, JSON responses, or test fixtures, and that redaction is consistent across HTTP and Connect surfaces.
4. **SSRF and outbound safety**: re-check base URL/host validation at config time, redirect time, and dial time for any preview that issues outbound calls; cover literal IPs, IPv6, link-local, DNS rebinding, and custom transports.
5. **Resource bounds and DoS**: verify request body size limits, response decoding caps, total in-flight previews per caller, and timeouts so a slow or huge upstream cannot starve the bootstrap service.
6. **Forward-compat with decode/live/audit PRs**: ensure preview event shape, event_id, and raw payload envelopes remain stable for the decoded preview, live GitHub, and audit-family follow-ups.
7. **Failure-mode classification**: confirm transient upstream/network/DNS failures stay retriable server errors while caller-fixable config errors become `InvalidArgument` consistently across all preview paths.
